### ISSUE TITLE
Pageserver didn't shutdown on signal

### ISSUE THREAD

**Issue by LizardWizzard** [zenithdb/zenith#1036]
Backtrace:
```
(gdb) thread apply all bt

Thread 2 (Thread 0x7f30ba50a700 (LWP 5877)):
#0  0x00007f30be991ddf in accept4 (fd=5, addr=..., addr_len=0x7f30ba5098ec, flags=524288) at ../sysdeps/unix/sysv/linux/accept4.c:32
#1  0x000055d66eba3439 in std::sys::unix::net::Socket::accept::{{closure}} () at library/std/src/sys/unix/net.rs:225
#2  std::sys::unix::cvt_r () at library/std/src/sys/unix/mod.rs:212
#3  std::sys::unix::net::Socket::accept () at library/std/src/sys/unix/net.rs:225
#4  std::sys_common::net::TcpListener::accept () at library/std/src/sys_common/net.rs:407
#5  std::net::tcp::TcpListener::accept () at library/std/src/net/tcp.rs:810
#6  0x000055d66e4ea563 in pageserver::page_service::thread_main (conf=0x55d670441aa0, auth=..., listener=..., auth_type=zenith_utils::postgres_backend::AuthType::Trust)
    at pageserver/src/page_service.rs:194
#7  0x000055d66e3925db in pageserver::start_pageserver::{{closure}} () at pageserver/src/bin/pageserver.rs:603
#8  std::sys_common::backtrace::__rust_begin_short_backtrace (f=...) at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/sys_common/backtrace.rs:125
#9  0x000055d66e392fd8 in std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}} () at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/thread/mod.rs:481
#10 <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once (self=..., _args=<optimized out>)
    at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/panic/unwind_safe.rs:271
#11 std::panicking::try::do_call (data=<optimized out>) at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panicking.rs:403
#12 std::panicking::try (f=...) at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panicking.rs:367
#13 std::panic::catch_unwind (f=...) at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/panic.rs:129
#14 std::thread::Builder::spawn_unchecked::{{closure}} () at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/thread/mod.rs:480
#15 core::ops::function::FnOnce::call_once{{vtable-shim}} () at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/ops/function.rs:227
#16 0x000055d66ebb1633 in <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/alloc/src/boxed.rs:1636
#17 <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/alloc/src/boxed.rs:1636
#18 std::sys::unix::thread::Thread::new::thread_start () at library/std/src/sys/unix/thread.rs:106
#19 0x00007f30bebe7fa3 in start_thread (arg=<optimized out>) at pthread_create.c:486
#20 0x00007f30be9904cf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

Thread 1 (Thread 0x7f30be892a40 (LWP 5871)):
#0  0x00007f30bebe9495 in __GI___pthread_timedjoin_ex (threadid=139847261005568, thread_return=0x0, abstime=0x0, block=<optimized out>) at pthread_join_common.c:89
#1  0x000055d66ebb170d in std::sys::unix::thread::Thread::join () at library/std/src/sys/unix/thread.rs:242
#2  0x000055d66e3927d2 in std::thread::JoinInner<T>::join (self=0x7fff03273ab8) at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/thread/mod.rs:1271
#3  std::thread::JoinHandle<T>::join (self=...) at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/std/src/thread/mod.rs:1403
#4  0x000055d66e39ea0c in pageserver::start_pageserver::{{closure}} (signal=<optimized out>) at pageserver/src/bin/pageserver.rs:627
#5  zenith_utils::signals::ShutdownSignals::handle (self=..., handler=...) at /zenith/zenith_utils/src/signals.rs:54
#6  0x000055d66e3af553 in pageserver::start_pageserver (conf=0x55d670441aa0) at pageserver/src/bin/pageserver.rs:607
#7  pageserver::main () at pageserver/src/bin/pageserver.rs:511
(gdb) 
```

**Comment by SomeoneToIgnore** [comment]
According to the dump, it's stuck on 
![image](https://user-images.githubusercontent.com/2690773/146757312-bc5116aa-5be3-460d-ac6e-5a5fbd05e8b4.png)


https://github.com/zenithdb/zenith/pull/787 attempts to address this and has a few comments about feasible alternatives.

**Comment by SomeoneToIgnore** [comment]
I've chased my ideas from https://github.com/zenithdb/zenith/pull/787#issuecomment-993355999 and checked the other crate. It turned to be a nice rust api around low level API that we already use in our repo:

https://github.com/zenithdb/zenith/blob/b5324707928502d7a99407a71f587d8141e5b81b/zenith_utils/src/tcp_listener.rs#L13

First, I've tried setting timeouts (both read and write), it was possible via "low-level" way and the other lib way, but neither helped to fix the "sigint single pageserver" blocking issue.

There's an interesting method in the other lib, that I've not found among the `setsockopt` options:
https://docs.rs/socket2/0.4.2/socket2/struct.Socket.html#method.set_nonblocking

If I use that, when I sigint the pageserver, it stops immediately, but it panics:
```
thread 'main' panicked at 'thread exited with an error: Resource temporarily unavailable (os error 35)', pageserver/src/bin/pageserver.rs:630:22
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I'll check the panic more in detail, but feels like it could work for us, if we could suppress it?
I've pushed a branch to see how our tests react on the nonblocking sockets: https://github.com/zenithdb/zenith/compare/kb/timed-accept?expand=1

**Comment by SomeoneToIgnore** [comment]
I've also decided to try out a simple experiment with `tokio`: I assumed, that we'll be stuck on some transition case where some of the listeners would be needed to be async, and some (such as pg connection) will stay sync for a while.

So, I've composed a small PR that takes `zenith_utils::tcp_listener::bind` and uses `tokio::net::TcpListener` to bind with `into_std()` to convert it back to `std::net::TcpListener` after binding.

https://github.com/zenithdb/zenith/compare/kb/async-accept?expand=1

That replicates similar behavior to my previous comment:
* "sigint single pageserver" test interrupts normally now
* tests are broken with `Connection refused (os error 61)`

So, it feels that either we have to move everything to async at once, or we'll have issues with whatever `std::net` primitives we get out of `tokio`.
